decisions: improving process (#4639)
ElektraInitiative/libelektra: doc/decisions/STEPS.md

# Steps

This document describes all steps a decision can run through.

```mermaid
flowchart LR
  s((Start)) --> Drafts --> In_Discussion --> In_Progress -- merge --> Decided
    -- merge --> Partially_Implemented --> Implemented

  %% Shortcuts:
  s --> Decided
  Drafts -- merge --> Decided
  In_Discussion -- merge --> Decided
  Decided -- merge --> Implemented

> ```

Reviewer: This needs a bit more explanation. What does the "merge" on the edge mean?

I guess, what you want to say is:

1. The first PR for a decision can be created as "Drafts", "In Discussion", "In Progress" or "Decided"
2. A decision PR that wants to merge a decision as "Decided" must:
    1. Already start with the decision as "Decided"
    2. Already have all the considered alternatives fully explored
    3. Only contain discussions/reviews about deciding between one of the considered alternatives
    4. If there is any significant change in the PR (i.e. information changes, not just typos) then the PR cannot be merged as "Decided". The decision must be moved back to "In Progress"
3. After a decision is merged as "Decided", it can be moved to "Partially Implemented" or "Implemented" by any PR.

If that is correct, maybe add this text or something similar in addition to the diagram.

Author: Thanks for the summary, I added it in bbe44da7c9d821ece2aecc2a1c400061302cf9d6

Furthermore I wrote:
> The label `merge` on the edges mean that a decision PR must be merged before it can target the next step.

- it clarifies relation to other problems.

Decisions will have much smoother further steps if this step is done carefully without prejudice.
It is especially important that one shouldn't have a fixed mind-set about a preferred solution from the beginning.

> Everyone must agree that the problem exists and is worth solving so that a decision PR in "Problem Clear" step can be merged.
> A problem is clear if everyone would be able to describe an experiment or test case that shows if a solution fixes the problem.

## Alternatives Clear

> This step is recommended if it is not yet clear to the core developers which solution is the best.

Here you must ensure:

- problem, constraint and assumptions are well-explained and sound
- consistency with other decisions
- links from/to related decisions are created
- there are several considered alternatives, each with rationale and implication
- "Decision", "Rationale" and "Implications" are **not** yet filled out if there are people arguing for different options

Here the decision should not only have one decision but should describe several solutions.
For each solution a proposal, rationale and optionally implications should be given.

> > The alternatives are clear if all reviewers understand the given alternatives and no reviewers can come up with new alternatives.

Reviewer: "no reviewers can come up with new alternatives" is a very bad criteria IMO. You can (almost) always come up with more alternatives. It also necessarily leads to long lists of alternatives that were never seriously considered because they are obviously flawed. But somebody had the idea, so we have to list them according to this definition...

Author: I agree, f832f42104bef6ada4da94c061cf698d928a19ff

```diff
-> and no reviewers can come up with new alternatives.
+> and no reviewer can come up with better alternatives.
```

> No agreement about anything is needed to merge decisions in this stage.

## Problem Clear

This step is very important:

- it clarifies the importance of the problem, answering:
  Why should we put our time and energy in this problem and not in another problem.
- it clarifies the scope of the decision.
- it clarifies relation to other problems.

Decisions will have much smoother further steps if this step is done carefully without prejudice.
It is especially important that one shouldn't have a fixed mind-set about a preferred solution from the beginning.

> Everyone must agree that the problem exists and is worth solving so that a decision PR in "Problem Clear" step can be merged.
> A problem is clear if everyone would be able to describe an experiment or test case that shows if a solution fixes the problem.

## Solutions Clear

> This step is recommended if it is not yet clear to the core developers which solution is the best.

Here you must ensure:

- problem, constraint and assumptions are well-explained and sound
- links from/to related decisions are created
- there are several solutions described, each with rationale and implication

> - "Decision", "Rationale" and "Implications" are **not** yet filled out if there are people arguing for different options

Reviewer: ```suggestion
- "Decision", "Rationale" and "Implications" are **not** yet filled out
```

If people don't argue for different options, this would be completely the same as "In Review". So better just say "not filled out at all" here.

Here the decision should not only have one decision but should describe several solutions.
For each solution a proposal, rationale and optionally implications should be given.

> The solutions are clear if all reviewers understand the given solutions and no reviewer can come up with better solutions.
> The solution space is clear.
> I.e. the trade-offs, combinations and pros/cons of the solutions are explored.
> Decision author and reviewers are satisfied that every useful solution is present in the decision.

## In Review

Now it is allowed to have the decision from the previous round in the "Decision" section.
In this step, the last details of the chosen decision get polished:

- consistency with other decisions
- links from/to related decisions checked

> - "Decision", "Rationale" and "Implications" are fully filled out

Reviewer: IMO this is just describes the state of an **open/unmerged** PR that wants to move/add a decision file to `doc/decisions/X_decided`. I don't really see a reason to ever merge a file in this state, _without_ also having the reviews and committing to the decision.

To me merging a fully filled out decision file in `doc/decisions/Y_in_review` seems like a way of creating a bias towards the decision the author wants. If the reviews already agree, then there is no need for this step. If reviews disagree merging the "Decision" section anyway seems kinda wrong.

Author: > merge a file in this state, without also having the reviews and committing to the decision.

There might be plenty of reasons, like one of the reviewers is not available anymore.

> seems like a way of creating a bias towards the decision the author wants

Yes but in this case it is allowed because there were already (hopefully unbiased) discussions showing this clear tendency.

> > Without merges in between, this is the last step reachable for a decision PR that started in "Drafts".

Reviewer: IMO creating a PR that introduces a new decision file and adds it directly to `doc/decisions/X_decided` should be fine. Since we need the reviews anyway, such a PR only makes sense, if the decision is only documenting status quo, consequences of other decisions, etc. Basically only decisions were there is only one realistic option.

Author: If the decisions are only documentation is does not really matter.

# Steps

This document describes all steps a decision can run through.

```mermaid
flowchart LR
    s((Start)) --> Drafts --> In_Discussion --> In_Progress --> Decided --> Partially_Implemented --> Implemented

    %% Shortcuts:
    s --> Decided
    Decided --> Implemented

    %% Dead ends:
    s --x Rejected
    s --o Delayed
    Drafts --x Rejected
    Drafts --o Delayed
    In_Discussion --x Rejected
    In_Discussion --o Delayed
    In_Progress --x Rejected
    In_Progress --o Delayed
```

> ## Drafts

Reviewer: ```suggestion
The first PR for a decision must **always** create the decision in the "Draft" state at first.
If during the reviews of this PR it becomes clear that the decision is further along, can be moved to "In Discussion" or "Decided" before the merge.

## Drafts
```

IMO this is a better process, because it makes clear how the first PR should be created. Otherwise the decision author needs to guess what the correct initial state of the decision is.

> > This step is recommended if the problem is not yet clear.

Reviewer: "Clear" to whom? In most cases the problem will be clear to the decision author.

Right now it's not at all clear when this step should be taken. With the extra sentence I suggested above, it would be clearer IMO. A decision always starts a "Draft". It can be merged as a "Draft", but it can also progress to "In Discussion" or even "Decided" before the merge.

- a link from [README.md](../README.md) from the "Drafts" section to this decision.
- optional backlinks from related decisions.

> Everyone must agree that the problem exists so that a decision PR in "Drafts" step can be merged.
> At least the problem must be clear to everyone involved before the decision can leave the "Drafts" step.
> It must be so clear that everyone would be able to describe a test case that shows if a solution fixes the problem.

## In Discussion

> This step is recommended if it is not yet clear which solution is the best.

Here you must ensure:

- problem, constraint and assumptions are well-explained and sound
- consistency with other decisions
- links from/to related decisions are created
- there are several considered alternatives, each with rationale and implication
- "Decision", "Rationale" and "Implications" are **not** yet filled out if there are people arguing for different options

Here the decision should not only have one decision but should describe several solutions.
For each solution a proposal, rationale and optionally implications should be given.

> ## In Progress

Reviewer: How is this different from "In Discussion" when would we use this?

IMO this step should be removed.

- You must include all further alternative proposals made in the "Considered Alternatives" section.
- Now it is allowed to have the decision from the previous round in the "Decision" section.

## Decided

> This step is mandatory.

- "Decision", "Rationale" and "Implications" are now filled out and fixed according to the reviews
- decisions of this step usually already have an implementation PR

> In this step, decision PRs only modify a _single_ decision.
> Only exceptions like backlinks from other decisions are allowed.

## Partially Implemented

> 

Reviewer: ```suggestion

> This step is optionally used when a decision cannot be implemented in a single PR.

```

> At least the problem must be clear to everyone involved before the decision can leave the "Drafts" step.
> It must be so clear that everyone would be able to describe a test case that shows if a solution fixes the problem.

## In Discussion

> This step is recommended if it is not yet clear to the core developers which solution is the best.

Here you must ensure:

- problem, constraint and assumptions are well-explained and sound
- consistency with other decisions
- links from/to related decisions are created
- there are several considered alternatives, each with rationale and implication
- "Decision", "Rationale" and "Implications" are **not** yet filled out if there are people arguing for different options

Here the decision should not only have one decision but should describe several solutions.
For each solution a proposal, rationale and optionally implications should be given.

## In Progress

- You must include all further alternative proposals made in the "Considered Alternatives" section.
- Now it is allowed to have the decision from the previous round in the "Decision" section.

> > For decisions "In Progress", unlike "In Discussion", also the solution space next to the problem is clear.

Reviewer: This doesn't help me one bit. I still don't understand the when "In Discussion" becomes "In Progress", but not "Decided" yet.

To me this feels like a valid step in the theoretical process, but doesn't fit with using PRs. IMO it sounds like a decision that has a decision (in the "Decision" section), but is not merged yet (because reviews aren't done) would be "In Progress". But that is a useless distinction. That only means you need to move the decision file, after the reviews are done.

When would we merge a decision as "In Progress"?

If I look at the current decision in [`doc/decisions/2_in_progress`](https://github.com/ElektraInitiative/libelektra/tree/master/doc/decisions/2_in_progress), the picture doesn't become much clearer. Many of them are "Decided" AFAIAC, some are even "Partially Implemented". Only `simplify_api.md` is still not decided IMO, but I'd say that's "In Discussion" because it is not clear what the exact API changes will be.

Author: > This doesn't help me one bit.

I elaborated the explanation in 5a330e8e85f47200d8dd965579f696855df00eac

> I still don't understand the when "In Discussion" becomes "In Progress", but not "Decided" yet.

Should be answered above. I feel like a merge point before we finally decide is cleaner.

> That only means you need to move the decision file, after the reviews are done.

For that we need good solutions anyway.

> When would we merge a decision as "In Progress"?

Always. It is a must, as it cannot be directly decided. (Unless it was directly target as decided for decisions where no discussion is needed. E.g. a decision needs another decision, split up of decisions etc..)

> Many of them are "Decided" AFAIAC, some are even "Partially Implemented".

These are left-overs before we had the decision process. I started cleanup in  #4690. The cleanup won't be finished with  #4690, though.

Reviewer: I think I get the "big picture" now see https://github.com/ElektraInitiative/libelektra/pull/4639#discussion_r1027069545. What I still don't quite understand is: Why we need both "In Discussion" and "In Progress"? Is it just a bookkeeping thing to know how close to a decision we are?

Maybe different names would be better (open to other suggestions):
- "Drafts" -> "Problem Draft" (because the problem is still a draft)
- "In Discussion" -> "Solution Draft" (problem is done, but considered alternatives, i.e. solutions, are still drafts)
- "In Progress" -> "In Review" (because theoretically everything is done, the review process just hasn't come to a decision yet)

If we want to keep the current names, IMO "In Discussion" and "In Progress" should be swapped. First it is "In Progress", i.e. we are working on possible solutions, _then_ it is "In Discussion", where we discuss only which solution is best.

Author: This is excellent input. I fully agree not only that the names are confusingly worded but trying to find better names also also something else became clear to me: Drafts should not have **any** requirements to be accepted. E.g. #4683 clearly is a well-written draft (clear problem, elaborate considered alternatives etc.). Just because of our definition of Drafts (that reviewers must agree on the problem), the PR is blocked. This hampers the possibility to collect proposals for problems that might occur in future. If we would already have had this draft in our drawer before thinking of COW, we would have probably implemented COW differently. But as COW is done now, the previous interesting decision now doesn't have a clear problem anymore. I agree that "Rejecting" such a decision is the wrong word. So I think it should be allowed to be merged as draft. If later another decision would benefit from having this decision, we can easily refer to it again.

I updated the decision process in 7691244f94ce63b266396546edfae56028bc3099
@kodebach Please review.

> > It is the highest of the steps reachable for a decision PR that started in "Drafts".

Reviewer: IMO that sentence is pretty confusing.

For one, I'd say "Decided" is the highest reachable step. Because I could create a PR with the decision as "Drafts" and move it trough the stages. Only when it reaches "Decided" do we require a merge, before we can move on.

Also the bit about "started in Drafts" is confusing. If I start the decision in "In Discussion" I could move beyond this step, or what?

Author: There is also the rule: "If during the reviews of this PR it becomes clear that the decision is further along, can be moved to "In Discussion" or "In Progress" before the merge." (which is the full list), so it is not possible to directly go to Decided from Drafts. See also updated Diagram: 674ec185e894c13cc22edb5d547352ef6ef70e8a

I changed the confusing sentence to: "Without merges in between, this is the last step reachable for a decision PR that started in "Drafts"." 570e48ab307995a791d7340d62cbbd23139e1331

```mermaid
flowchart LR
  s((Start)) --> Drafts --> Problem_Clear --> Alternatives_Clear --> In Review -- merge --> Decided
    -- merge --> Partially_Implemented --> Implemented

  %% Shortcuts:
  s --> Decided
  Drafts -- merge --> Decided
  Problem_Clear -- merge --> Decided
  Alternatives_Clear -- merge --> Decided
  Decided -- merge --> Implemented
```

> Additionally, decision that are not yet "Decided" can be become "Drafts", "Rejected" or "Postponed" at any point, e.g., if the decision author stops working on the decision.

Author: I changed can -> may. This is something rare, so I don't think we need to write rules for this. 00ddc456120d01b145a473e31904966875886ead

Here the decision should not only have one decision but should describe several solutions.
For each solution a proposal, rationale and optionally implications should be given.

## In Review

- You must include all further alternative proposals made in the "Considered Alternatives" section.
- Now it is allowed to have the decision from the previous round in the "Decision" section.

> For decisions "In Review" the solution space is clear.
> I.e. the trade-offs, combinations and pros/cons of the considered alternatives are explored.
> Decision author and reviewers are satisfied that every useful considered alternative is present in the decision.
> Without merges in between, this is the last step reachable for a decision PR that started in "Drafts".

## Decided

> This step is mandatory.
> I.e., there must be a dedicated decision PR that puts the decision into "Decided".

- "Decision", "Rationale" and "Implications" are now filled out and fixed according to the reviews
- decisions of this step usually already have an implementation PR

> Decisions that need an update, e.g. because assumptions changed, usually directly start with the step "Decided".

Reviewer: I think that needs to be phrase differently. IMO it should only start in "Decided", if the chosen solution stays the same and only details change. If the problem it self changes, or the chosen solution is no longer a good choice, then we should probably start in an earlier step.

Author: Changed usually -> sometimes. I don't think we should put rules on rare cases. e447fde4d313e45aafc416f7073a520845dc8dba

- The decision links to the new documentation.

> In this step, decision PRs only modify a _single_ decision.
> Here more exceptions are allowed, in particular documentation updates are okay.

## Postponed

This step is:

> - for decisions that are useful contributions but currently there is simply no urgency or immanent need for the proposal.

Reviewer: ```suggestion
- for decisions that would be useful contributions but there is currently nobody with the time to focus on the problem.
```

If there is a clear problem and there is a person who wants to work on it, why not let them? They may not want to or have the skill to work on other problems, so refusing their contribution would just be a waste.

- a graveyard for neglected decisions, e.g., where the decision authors focus on something more important instead.

> The purpose of this step is to have a clean "Drafts" folder.

## Rejected

Alternatively, decisions might be rejected (e.g. status quo wins).

> This step is for decisions that are agreed to not fit well to Elektra and never should be accepted.

Reviewer: ```suggestion
This step is for decisions for which a consensus exists that the decision should not be taken right now.
If circumstances change, the decision may be revisited.
```
